Infer 0.5 (new formula)

tonycosentini opened the pull request:
This adds a formula for Infer - a static analysis tool from Facebook.

bfontaine commented on Library/Formula/fb-infer.rb:
```diff
@@ -0,0 +1,56 @@
+class FbInfer < Formula
+  desc "Static analyzer for Java, C and Objective-C"
+  homepage "http://fbinfer.com/"
+
```
Please remove this empty line.

bfontaine commented on Library/Formula/fb-infer.rb:
```diff
@@ -0,0 +1,56 @@
+
+  url "https://github.com/facebook/infer/releases/download/v0.5.0/infer-osx-v0.5.0.tar.xz"
+  sha256 "6a8547ac0b75a5e2bbeccae2169e39f753a60adbcacb6c94599fd31343a71ce7"
+  revision 1
```
Please remove this `revision`.

bfontaine commented on Library/Formula/fb-infer.rb:
```diff
@@ -0,0 +1,56 @@
+      target_language = "clang"
+    end
+
+    mktemp("fb-infer-opam-root") do
```
Can’t you create a directory in the same tree, e.g. `mkdir "fb-infer-opam-root" do`?

tonycosentini replied:
It can't be in the same tree, because then Opam will fail :disappointed:

```
[ERROR] Cannot sync
        /private/tmp/fb-infer20160107-59822-kstm75/infer-osx-v0.5.0/ into
        /private/tmp/fb-infer20160107-59822-kstm75/infer-osx-v0.5.0/fb-infer-opam-root/system/packages.dev/infer:
        they overlap
[infer] /private/tmp/fb-infer20160107-59822-kstm75/infer-osx-v0.5.0/ unavailable
infer is now unpinned from path /private/tmp/fb-infer20160107-59822-kstm75/infer-osx-v0.5.0
```

UniqMartin replied:
Please have at the [`coccinelle`](https://github.com/Homebrew/homebrew/blob/master/Library/Formula/coccinelle.rb), [`fstar`](https://github.com/Homebrew/homebrew/blob/master/Library/Formula/fstar.rb), or [`zero-install`](https://github.com/Homebrew/homebrew/blob/master/Library/Formula/zero-install.rb) formula to see how to best work with `opam` in your `install` method. I'm fairly optimistic the same approach works with this formula, too. Thanks!

tonycosentini replied:
These packages don't pin a local package though so I don't think they can run in to the same problem.

UniqMartin replied:
Can you explain to someone (me) who is not familiar with `opam` how that makes a difference?

tonycosentini replied:
Sure -  I think this is mostly an issue with Infer using a very strange build process, but I'm also new to Opam.

Infer lists all of their Opam dependencies in their opam file:  https://github.com/facebook/infer/blob/master/opam#L16

This formula marks the current directory as an Opam library and then installs all the dependencies:
```
system "opam", "pin", "add", "--yes", "--no-action", "infer", "."
system "opam", "install", "--deps-only", "infer"
```

Unfortunately, you can't pin a directory that if the Opam root is that directory, is in a directory of that directory.

This is the main difference between the examples posted above, it appears they use a different approach to dependencies.

UniqMartin replied:
Thank you for the explanation! Unfortunately, you can't use `mktemp` to work around this. `mktemp` creates a temporary directory outside of the build directory, but the formula install is supposed to stay within the build directory and a few other hand-picked locations.

On some systems, the formula build process is sandboxed to prevent it from accessing arbitrary paths, hence using the `mktemp` approach will make it fail. I assume that's what happened with your most recent Travis build, though I'm not 100% sure.

I'm somewhat struggling with a good suggestion for how to solve your problem, though. Maybe you can relocate the contents of `buildpath` into a subdirectory and then create the `OPAMROOT` next to it? (Maybe another maintainer has a better idea.)

tonycosentini replied:
The subdirectory approach would work, do you know of any other formulas doing this?

UniqMartin replied:
Sadly, no. This should be simple in theory, but in practice you have to be careful only to move the files that were extracted from the archive and skip over the stuff created by Homebrew (there's a `.brew_home` directory in `buildpath` and it needs to stay where it is). I don't want to recommend this approach just yet.

Do you think there is a feasible way to tweak things on the source side so that we can work with the existing directory layout (and do things more similarly to the other formulae that use `opam`)?

tonycosentini replied:
I can try an approach similar to the other formulae.

tonycosentini replied:
Thanks for the helpful feedback!

bfontaine commented on Library/Formula/fb-infer.rb:
```diff
@@ -0,0 +1,56 @@
+    end
+
+    mktemp("fb-infer-opam-root") do
+      ENV["OPAMROOT"] = Pathname.new(Dir.pwd)
```
If you use `mkdir` above you’ll be able to use `buildpath/"fb-infer-opam-root"` or `buildpath/"build"` here instead of `Pathname.new(Dir.pwd)`.

bfontaine commented on Library/Formula/fb-infer.rb:
```diff
@@ -0,0 +1,56 @@
+  end
+
+  test do
+    system "infer"
```
Use `"#{bin}/infer"` instead.

bfontaine commented:
If you need an empty directory for opam you can use something like:
```rb
opamroot = buildpath/"build"
opamroot.mkpath
ENV["OPAMROOT"] = opamroot
```

bfontaine commented on Library/Formula/fb-infer.rb:
```diff
@@ -0,0 +1,56 @@
+  depends_on "opam" => :build
+
+  def install
+    ENV.deparallelize
```
Is upstream aware of this?

tonycosentini replied:
This actually isn't needed, somehow just got added when I was debugging the script. I'll remove it.

tonycosentini commented:
I'm not sure why this failed on 10.11 only with Travis, going to restart to see if it was some weird connectivity issue or something.

bfontaine commented:
@tonycosentini It’s ok for the Travis build to fail; only the `default` job is important for now.

tonycosentini commented:
Ignore this current revision, accidentally pushed when I shouldn't have.

tonycosentini commented:
@UniqMartin I just updated this to take the same approach as coccinelle does - seems to work great.

I'll check with the Infer folks if there is any way to streamline this, but I think this approach is fine for this release.

Thanks again for the feedback - very helpful.

UniqMartin commented on Library/Formula/fb-infer.rb:
```diff
@@ -0,0 +1,53 @@
+class FbInfer < Formula
```
Why did you decide to go with `fb-infer` as the formula name? The Website is <http://fbinfer.com/> and the repository name is just `infer`. Most existing formulae for Facebook's tools use just the repository name, so I'd suggest to use simply `infer`. (This would also make it match the name of the installed binary.)

bfontaine replied:
:+1: for `infer`; this is how upstream calls it. They probably used fbinfer.com because infer.com is already taken.

UniqMartin commented on Library/Formula/fb-infer.rb:
```diff
@@ -0,0 +1,53 @@
+  head "https://github.com/facebook/infer.git"
+
+  option "with-clang-only", "Installs Infer with only Clang support."
+  option "with-java-only", "Installs Infer with only Java support."
```
Please omit the full stop at the end of the option descriptions. We also tend to phrase these things `Build with <feature>` (no need to repeat the formula name here), so I'd suggest:

- `Build with C/Objective-C analyzer only` (this makes it also clearer than just stating `Clang`)
- `Build with Java analyzer only`

UniqMartin commented on Library/Formula/fb-infer.rb:
```diff
@@ -0,0 +1,53 @@
+  homepage "http://fbinfer.com/"
+  url "https://github.com/facebook/infer/releases/download/v0.5.0/infer-osx-v0.5.0.tar.xz"
+  sha256 "6a8547ac0b75a5e2bbeccae2169e39f753a60adbcacb6c94599fd31343a71ce7"
+  head "https://github.com/facebook/infer.git"
```
Just out of curiosity: Have you tested whether the `--HEAD` build succeeds at this point of time?

UniqMartin commented on Library/Formula/fb-infer.rb:
```diff
@@ -0,0 +1,53 @@
+  depends_on "opam" => :build
+
+  def install
+    target_language = "all"
```
Can you move this line directly before `if build.with?("java-only")`? I think this will make the grouping in the `install` method clearer.

UniqMartin commented on Library/Formula/fb-infer.rb:
```diff
@@ -0,0 +1,53 @@
+      target_language = "java"
+    elsif build.with?("clang-only")
+      target_language = "clang"
+    end
```
This is purely cosmetic, but why have you decided to test for `java-only` first where in all other instances `clang-only` comes first. (Feel free to ignore this, just wanted to point it out since the formula needs a few more tweaks anyway.)

UniqMartin commented on Library/Formula/fb-infer.rb:
```diff
@@ -0,0 +1,53 @@
+  end
+
+  test do
+    system "#{bin}/infer"
```
I suspect this is just printing usage instructions. Can you make this test do something more substantial than this? A good test case would be something that uses the actual functionality of `infer`, so probably a minimal Java and/or C program that causes `infer` to report an error.

UniqMartin commented:
> I just updated this to take the same approach as coccinelle does - seems to work great.

Thanks for following through a finding a much better solution! The Travis build is still failing, so there's probably something else going on. But you can ignore Travis if the `default` build succeeds.

I added a few more remarks now that the build itself is in a good shape. Please address these and then I think we're good to go.

tonycosentini commented:
OK, just pushed a new revision with:
* Renamed formula.
* Responded to cosmetic changes suggested.
* Improved test to actually run Infer and verify that it works.
* Removed support for installing head - there are some issues with the Facebook clang plug-in, I’m just going to remove this for now.

UniqMartin commented:
The CI isn't happy with your `get_target_platform` method and I'm neither (it's a bit over-engineered, no need for the constants either). To catch these types of failures earlier, you can run `brew audit --strict infer` locally, which will run a few checks on the formula.

I think I understand why you introduced this `get_target_platform` method, but to me it finally revealed that maybe the options are named sub-optimally. Here's what I suggest to call them:

- `without-clang`: “Build without C/Objective-C analyzer”
- `without-java`: “Build without Java analyzer”

This inverts the logic, but makes everything easier in this formula (and for people who look at its options). You would still perform the sanity check at the beginning:

```ruby
if build.without?("clang") && build.without?("java")
  odie "infer: --without-clang and --without-java are mutually exclusive"
end
```

You continue by inferring the target platform. More idiomatic Ruby code for this would be:

```ruby
target_platform = if build.without?("clang")
  "java"
elsif build.without?("java")
  "clang"
else
  "all"
end
```

A shorter (but more repetitive) alternative would be:

```ruby
target_platform = "all"
target_platform = "java" if build.without?("clang")
target_platform = "clang" if build.without?("java")
```

You can use that as you did previously in the invocation of `build-infer.sh`. The two parts of the `test do` block (by the way: awesome tests you added here; I wish every `test do` of a new formula looked like this) can now be guarded by either `if build.with?("clang")` or `if build.with?("java")`.

tonycosentini commented:
Hey - I just pushed a change to address the latest feedback.

Thanks again for the code review!

bfontaine commented on Library/Formula/infer.rb:
```diff
@@ -0,0 +1,116 @@
+  end
+
+  test do
+    if build.with?("clang")
```
You can now remove these `if build.with?` in the test because they’ll be true by default and we only test the default install.

mikemcquaid replied:
Tests can always assume the default options, just FYI.

UniqMartin replied:
I would prefer if we can keep this conditional, as I think a user should be able to build from source (when building with custom options) and the tests should still complete successfully. Call me crazy, but I actually run `brew test <formula>` locally sometimes after doing exactly that.

mikemcquaid replied:
In this case those are all default options.

DomT4 replied:
I'd really rather we didn't include anything that isn't the default. Testing non-default options opens a floodgate of testing possibilities that we're not likely to be able to cram back into the jar.

We don't build non-default options on CI so not building them to know whether they still work or not and then allowing users to test those non-default options seems like a recipe for user confusion and maintainer frustration.

UniqMartin replied:
I saw value in that, but I shall be overridden.

DomT4 replied:
Just think it'd be nice to be consistent, i.e. if we support testing non-default options we should CI build them as well, so we know they won't die horribly, etc.

tonycosentini replied:
What's the conscious?

mikemcquaid commented on Library/Formula/infer.rb:
```diff
@@ -0,0 +1,116 @@
+    system "./build-infer.sh", target_platform, "--yes"
+
+    libexec.install "facebook-clang-plugins"
+    libexec.install "infer/"
```
`libexec.install "facebook-clang-plugins", "infer"`

mikemcquaid commented:
Failing to bottle, looks like a bad symlink: http://bot.brew.sh/job/Homebrew%20Pull%20Requests/39731/version=el_capitan/testReport/junit/brew-test-bot/el_capitan/bottle_infer/

tonycosentini commented:
Just pushed a change that fixed the symlink issue (there's an unnecessary symlink pointing to a file not copied, so I just removed it with the script) and reverted back to the old test.

UniqMartin commented on Library/Formula/infer.rb:
```diff
@@ -0,0 +1,54 @@
+    if build.without?("clang") && build.without?("java")
+      odie "infer: --without-clang and --without-java are mutually exclusive"
+    end
+    opamroot = buildpath/"build"
```
Only cosmetic, but please add an empty line between the check and the beginning of the `opam` stuff.

UniqMartin commented on Library/Formula/infer.rb:
```diff
@@ -0,0 +1,54 @@
+
+    if build.with?("clang")
+      libexec.install "facebook-clang-plugins"
+    end
```
This can be shortened to `libexec.install "facebook-clang-plugins" if build.with?("clang")`.

UniqMartin commented on Library/Formula/infer.rb:
```diff
@@ -0,0 +1,54 @@
+  end
+
+  test do
+    system "#{bin}/infer", "-v"
```
Please bring back the great tests that you added in a previous iteration. And please run them for both languages unconditionally, i.e., without the `if build.with?("clang")` etc. I suggested before my fellow maintainers asked to do otherwise. (The consensus I wasn't aware of is to assume a default installation in `test do`. Sorry for misguiding you. I wasn't aware of that.)

tonycosentini replied:
Ah gotcha, will do.

tonycosentini commented:
Can anyone clarify why the 10.11 build consistently fails on Travis, but not on Jenkins?

UniqMartin commented:
> Can anyone clarify why the 10.11 build consistently fails on Travis, but not on Jenkins?

I tried to understand the Travis failure, but didn't succeed so far. But since Jenkins is fine with is, so are we for the time being. Thanks for the amount of work you put into this!

Merged in 1a77218b8f236fcfa1f0654fee2d64f29e43faab. Thank you for this contribution to Homebrew, @tonycosentini! :tada:

tonycosentini commented:
Thanks for the help and review! I appreciate the time spent improving this.
